src/Components/Server/src/Circuits/ServerLazyAssemblyLoader.cs

```diff
@@ -0,0 +1,19 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Server
+{
+    public class ServerLazyAssemblyLoader : ILazyLoader
```

Comment:
Do we definitely need this? I'm concerned developers will get really confused if there's any equivalent concept to lazy loading in server-side Blazor. If possible this should only be a thing in WebAssembly world.

Reply:
We need something like this to support prerendering, but I'm not sure it should live here. We might need to provide an implementation on MVC.

I think it will be worth to spend some time reworking how we handle the dependencies so that we can have abstractions that live on webassembly and that can be initialized on a server context without MVC having to know about each one of the abstractions.

Reply:
I guess the abstraction could go into `M.A.C.Web`. That's already referenced by `M.A.C.WebAssembly`. It's *not* currently referenced by `M.A.C.Server` but it seems reasonable to do so, since "server" implies you're working with web technologies.

Reply:
As per [this discussion](https://github.com/dotnet/aspnetcore/pull/23290#discussion_r450939212), hopefully no such abstraction is needed in the end.

Reply:
Javier is correct about having this to support pre-rendering.

I was mimicking the pattern that existed for the NavigationManager and NavigationInterceptor since those were the closest conceptual to the loader. Although, notably, following that pattern to the letter would mean that `ILazyLoader` should be an abstract class as discussed in the other thread.

> I think it will be worth to spend some time reworking how we handle the dependencies so that we can have abstractions that live on webassembly and that can be initialized on a server context without MVC having to know about each one of the abstractions.

Can you clarify what you mean for this? I'm interpreting "handle the dependencies" to mean how we handle the app's dependency DLLs.

> I guess the abstraction could go into M.A.C.Web. That's already referenced by M.A.C.WebAssembly. It's not currently referenced by M.A.C.Server but it seems reasonable to do so, since "server" implies you're working with web technologies.

It didn't make sense to me to put it in `M.A.C.Web` because that namespace seemed to deal heavily with DOM interaction and the lazy loader was more component lifecycle.

Is there a reason we don't want it in `M.A.C`?

Reply:
If we’re willing to treat wasm+lazyloading+prerendering as an uncommon case, there may be a pretty simple solution. Like Javier says, we can require devs to register a service to make this work on the server. 

That service could be the actual WebAssembly loader service, which internally knows how to behave usefully when running on the server (basically, just no-op). Then no deeper abstraction is needed at all. We could still change this nonbreakingly in the future by adding an abstraction under it and creating a server flavour, but hopefully that would never be needed. 

Does that sound feasible?

Reply:
Yep, in fact, that's what we ended up with here.